src/libstore/store-api.cc

```diff
@@ -191,12 +191,13 @@ StorePath Store::makeFixedOutputPath(std::string_view name, const FixedOutputInf
         return makeStorePath(makeType(*this, "source", info.references), info.hash, name);
     } else {
         assert(info.references.size() == 0);
-        return makeStorePath("output:out",
-            hashString(htSHA256,
-                "fixed:out:"
+        // make a unique digest based on the parameters for creating this store object
+        // FIXME: this could use an abstraction to encapsulate the specifics of this construction
```

Comment:
What does this mean? I would think `FixedOutputInfo` is the abstraction

Reply:
`FixedOutputInfo` could have a `to_string()` method that tucks away these particularities. Also it would be good to encapsulate the magic strings that define the various store path types.

Reply:
@fricklerhandwerk So unfortunately Nix is not entirely consistent about its magic strings --- the magic strings used in derivation outputs' ATerm are not precisely the same as the magic strings used when calculating store paths.

This makes me think that the magic strings are not a property of the content addressing types themselves, but a property of *the derivation encoding* itself and *the store path encoding* itself. In other words the code is in the right spot.

(C.f. how all the worker protocol serialization code lives in one place, and not where the data types being serialized are defined. Serializations are not intrinsic properties of the thing being serialized.)

If anything I think I've gone gone too far moving magic string stuff to `content-address.hh`, because that violates the principle defined above.

Reply:
> In other words the code is in the right spot.

I'd still suggest having wrappers to these magic strings, and let them live "in the right spot", wherever you think that is.

Reply:
```suggestion
        // make a unique digest based on the parameters for creating this store object
```